src/rules/noTrailingWhitespaceRule.ts

```diff
@@ -72,53 +73,49 @@ export class Rule extends Lint.Rules.AbstractRule {
 }
 
 function walk(ctx: Lint.WalkContext<IgnoreOption>) {
-    let lastSeenWasWhitespace = false;
-    let lastSeenWhitespacePosition = 0;
-    utils.forEachTokenWithTrivia(ctx.sourceFile, (fullText, kind, range) => {
-        if (kind === ts.SyntaxKind.NewLineTrivia || kind === ts.SyntaxKind.EndOfFileToken) {
-            if (lastSeenWasWhitespace) {
-                reportFailure(ctx, lastSeenWhitespacePosition, range.pos);
+    const possibleFailures: ts.TextRange[] = [];
+    const sourceFile = ctx.sourceFile;
+    const text = sourceFile.text;
+    for (const range of getLineRanges(sourceFile)) {
```

Comment:
Your solution looks nicer at first glance. It seems to include `\r` in the range of the failure, though.

I just added a new property `contentLength` to `getLineRanges` which has already stripped `\r?\n`. The regex is now even simpler.